contracts/TaskBook.sol

```diff
@@ -5,83 +5,158 @@ import './AccountManager.sol';
+		t.state = 1;
+		log0(randomBitsHash);
+		SolverSelected(taskID, msg.sender, t.taskData, t.minDeposit);
 		return true;
 	}
 
 	//Selected solver submits a solution to the exchange
-	function commitSolution(uint taskID, bytes32 solutionHash) returns (bool) {
-		Task t = tasks[taskID];
+	//1->2
+	function commitSolution(uint taskID, bytes32 solutionHash0, bytes32 solutionHash1) returns (bool) {
+		Task storage t = tasks[taskID];
 		require(t.selectedSolver == msg.sender);
-		solutions[msg.sender][taskID] = solutionHash;
-		SolutionCommitted(taskID, t.minDeposit, t.taskData, msg.sender);
+		require(t.state == 1);
+		Solution storage s;
+		s.solutionHash0 = solutionHash0;
+		s.solutionHash1 = solutionHash1;
+		solutions[taskID] = s;
+		t.state = 2;
+		SolutionsCommitted(taskID, t.minDeposit, t.taskData, msg.sender);
 		return true;
 	}
 
 	//Verifier submits a challenge to the solution provided for a task
-	function commitChallenge(uint taskID, uint minDeposit, address solver, bytes32 intentHash) returns (bool) {
+	//Verifiers can call this until task giver changes state or timeout
+	function commitChallenge(uint taskID, uint minDeposit, bytes32 intentHash) returns (bool) {
 		require(balances[msg.sender] >= minDeposit);
-		require(!(solutions[solver][taskID] == 0x0));
-		Task t = tasks[taskID];
-		require(t.numChallengers < maxChallengers);
-		verifierIntent[msg.sender][taskID] = intentHash;
-		t.challengers.push(msg.sender);
-		log0(sha3(msg.sender));
+		Task storage t = tasks[taskID];
+		require(t.state == 2);
+		t.challenges[msg.sender] = intentHash;
 		return true;
 	}
-}
\ No newline at end of file
+
+	//Verifiers can call this until task giver changes state or timeout
+	function revealIntent(uint taskID, uint intent) returns (bool) {
+		require(tasks[taskID].challenges[msg.sender] == sha3(intent));
+		require(tasks[taskID].state == 3);
+		if(intent % 2 == 0) {//Intent determines which solution the verifier is betting is deemed incorrect
+			solutions[taskID].solution0Challengers.push(msg.sender);
+		}else{
+			solutions[taskID].solution1Challengers.push(msg.sender);
+		}
+		delete tasks[taskID].challenges[msg.sender];
+		return true;
+	}
+
+	//4->5
+	function revealSolution(uint taskID, bool solution0Correct, uint originalRandomBits) returns (bool) {
+		require(solverRandomBitsHash[msg.sender][taskID] == sha3(originalRandomBits));
+		require(tasks[taskID].state == 4);
+		require(tasks[taskID].selectedSolver == msg.sender);
+		solutions[taskID].solution0Correct = solution0Correct;
+		tasks[taskID].state = 5;
+		SolutionRevealed(taskID, originalRandomBits);
+		return true;
+	}
+
+	//5->6
+	//This assumes that the verification game will state who gets paid
+	function verifySolution(uint taskID, uint randomBits) returns (bool) {
+		require(tasks[taskID].state == 5);
+		require(tasks[taskID].owner == msg.sender);
+		tasks[taskID].state = 6;
+		if(uint(sha3(randomBits, block.blockhash(block.number))) < forcedErrorThreshold) {//Forced error
```

Comment:
`block.blockhash(block.number)` is 0, perhaps it would be best to store the block number when the random bits are commited, I think that way there is least amount of gaming possible.